File: Pillars/Pillars/Models/CoachPreferences.swift
//
//  CoachPreferences.swift
//  Pillars
//
//  Coach preferences that personalize AI coaching
//

import Foundation

struct CoachPreferences: Codable, Equatable {
    var id: String?
    var userId: String?
    
    // Communication Style
    var communicationStyle: CommunicationStyle
    var tone: CoachingTone
    
    // Check-in Preferences
    var checkInFrequency: CheckInFrequency
    var preferredTime: PreferredTime
    
    // Coaching Focus
    var focusAreas: [FocusArea]
    
    // Additional Preferences
    var useEmojis: Bool
    var messageLength: MessageLength
    var proactiveCheckIns: Bool
    
    var createdAt: Date?
    var updatedAt: Date?
    
    // MARK: - Default
    static var `default`: CoachPreferences {
        CoachPreferences(
            communicationStyle: .balanced,
            tone: .supportive,
            checkInFrequency: .whenIReachOut,
            preferredTime: .anytime,
            focusAreas: [.goals, .habits],
            useEmojis: true,
            messageLength: .concise,
            proactiveCheckIns: false
        )
    }
    
    // MARK: - Generate Prompt Instructions
    func toPromptInstructions() -> String {
        var instructions: [String] = []
        
        // Communication style
        switch communicationStyle {
        case .direct:
            instructions.append("Be direct and to-the-point. Skip pleasantries and get straight to actionable advice.")
        case .gentle:
            instructions.append("Be warm, gentle, and encouraging. Take time to acknowledge feelings before offering guidance.")
        case .balanced:
            instructions.append("Balance warmth with directness. Be friendly but focused.")
        }
        
        // Tone
        switch tone {
        case .motivational:
            instructions.append("Use a motivational, energizing tone. Celebrate wins and push for growth.")
        case .analytical:
            instructions.append("Be analytical and logical. Focus on data, patterns, and systematic approaches.")
        case .supportive:
            instructions.append("Be supportive and understanding. Prioritize emotional support and validation.")
        case .challenging:
            instructions.append("Be challenging and push back constructively. Ask tough questions to promote growth.")
        }
        
        // Focus areas
        if !focusAreas.isEmpty {
            let areas = focusAreas.map { $0.rawValue }.joined(separator: ", ")
            instructions.append("Focus coaching on: \(areas).")
        }
        
        // Message length
        switch messageLength {
        case .brief:
            instructions.append("Keep responses very brief (1-2 sentences max).")
        case .concise:
            instructions.append("Keep responses concise (2-4 sentences).")
        case .detailed:
            instructions.append("Provide detailed responses with examples when helpful.")
        }
        
        // Emojis
        if useEmojis {
            instructions.append("Feel free to use emojis occasionally to add warmth.")
        } else {
            instructions.append("Avoid using emojis.")
        }
        
        return instructions.joined(separator: " ")
    }
}

// MARK: - Enums

enum CommunicationStyle: String, Codable, CaseIterable {
    case direct = "Direct"
    case gentle = "Gentle"
    case balanced = "Balanced"
    
    var description: String {
        switch self {
        case .direct: return "Straight to the point, no fluff"
        case .gentle: return "Warm and encouraging"
        case .balanced: return "Mix of warmth and directness"
        }
    }
    
    var icon: String {
        switch self {
        case .direct: return "bolt.fill"
        case .gentle: return "heart.fill"
        case .balanced: return "scale.3d"
        }
    }
}

enum CoachingTone: String, Codable, CaseIterable {
    case motivational = "Motivational"
    case analytical = "Analytical"
    case supportive = "Supportive"
    case challenging = "Challenging"
    
    var description: String {
        switch self {
        case .motivational: return "Energizing and encouraging"
        case .analytical: return "Logical and data-driven"
        case .supportive: return "Understanding and validating"
        case .challenging: return "Pushes you to grow"
        }
    }
    
    var icon: String {
        switch self {
        case .motivational: return "flame.fill"
        case .analytical: return "chart.bar.fill"
        case .supportive: return "hand.raised.fill"
        case .challenging: return "figure.climbing"
        }
    }
}

enum CheckInFrequency: String, Codable, CaseIterable {
    case daily = "Daily"
    case weekly = "Weekly"
    case whenIReachOut = "When I reach out"
    
    var description: String {
        switch self {
        case .daily: return "Check in with me every day"
        case .weekly: return "Weekly check-ins are enough"
        case .whenIReachOut: return "Only when I start a conversation"
        }
    }
}

enum PreferredTime: String, Codable, CaseIterable {
    case morning = "Morning"
    case afternoon = "Afternoon"
    case evening = "Evening"
    case anytime = "Anytime"
    
    var icon: String {
        switch self {
        case .morning: return "sunrise.fill"
        case .afternoon: return "sun.max.fill"
        case .evening: return "moon.fill"
        case .anytime: return "clock.fill"
        }
    }
}

enum FocusArea: String, Codable, CaseIterable {
    case goals = "Goals"
    case habits = "Habits"
    case emotions = "Emotions"
    case relationships = "Relationships"
    case career = "Career"
    case health = "Health"
    case mindfulness = "Mindfulness"
    
    var icon: String {
        switch self {
        case .goals: return "target"
        case .habits: return "repeat"
        case .emotions: return "heart.fill"
        case .relationships: return "person.2.fill"
        case .career: return "briefcase.fill"
        case .health: return "figure.run"
        case .mindfulness: return "brain.head.profile"
        }
    }
}

enum MessageLength: String, Codable, CaseIterable {
    case brief = "Brief"
    case concise = "Concise"
    case detailed = "Detailed"
    
    var description: String {
        switch self {
        case .brief: return "1-2 sentences"
        case .concise: return "2-4 sentences"
        case .detailed: return "Longer, with examples"
        }
    }
}

File: backend/src/routes/coach-preferences.js
const express = require('express');
const router = express.Router();
const { db, FieldValue } = require('../config/firebase');
const { verifyToken } = require('../middleware/auth');

router.use(verifyToken);

// GET - Get user's coach preferences
router.get('/', async (req, res) => {
  try {
    const doc = await db.collection('coachPreferences').doc(req.user.uid).get();
    
    if (!doc.exists) {
      // Return defaults if no preferences exist
      return res.json({
        id: req.user.uid,
        userId: req.user.uid,
        communicationStyle: 'Balanced',
        tone: 'Supportive',
        checkInFrequency: 'When I reach out',
        preferredTime: 'Anytime',
        focusAreas: ['Goals', 'Habits'],
        useEmojis: true,
        messageLength: 'Concise',
        proactiveCheckIns: false
      });
    }
    
    res.json({ id: doc.id, ...doc.data() });
  } catch (error) {
    console.error('Error getting coach preferences:', error);
    res.status(500).json({ error: error.message });
  }
});

// PUT - Update coach preferences
router.put('/', async (req, res) => {
  try {
    const {
      communicationStyle,
      tone,
      checkInFrequency,
      preferredTime,
      focusAreas,
      useEmojis,
      messageLength,
      proactiveCheckIns
    } = req.body;
    
    const preferences = {
      userId: req.user.uid,
      communicationStyle: communicationStyle || 'Balanced',
      tone: tone || 'Supportive',
      checkInFrequency: checkInFrequency || 'When I reach out',
      preferredTime: preferredTime || 'Anytime',
      focusAreas: focusAreas || ['Goals', 'Habits'],
      useEmojis: useEmojis !== undefined ? useEmojis : true,
      messageLength: messageLength || 'Concise',
      proactiveCheckIns: proactiveCheckIns || false,
      updatedAt: FieldValue.serverTimestamp()
    };
    
    // Check if document exists
    const docRef = db.collection('coachPreferences').doc(req.user.uid);
    const doc = await docRef.get();
    
    if (!doc.exists) {
      preferences.createdAt = FieldValue.serverTimestamp();
    }
    
    await docRef.set(preferences, { merge: true });
    
    res.json({ id: req.user.uid, ...preferences });
  } catch (error) {
    console.error('Error saving coach preferences:', error);
    res.status(500).json({ error: error.message });
  }
});

// Helper function to generate prompt instructions from preferences
function generatePromptInstructions(preferences) {
  const instructions = [];
  
  // Communication style
  switch (preferences.communicationStyle) {
    case 'Direct':
      instructions.push('Be direct and to-the-point. Skip pleasantries and get straight to actionable advice.');
      break;
    case 'Gentle':
      instructions.push('Be warm, gentle, and encouraging. Take time to acknowledge feelings before offering guidance.');
      break;
    case 'Balanced':
    default:
      instructions.push('Balance warmth with directness. Be friendly but focused.');
  }
  
  // Tone
  switch (preferences.tone) {
    case 'Motivational':
      instructions.push('Use a motivational, energizing tone. Celebrate wins and push for growth.');
      break;
    case 'Analytical':
      instructions.push('Be analytical and logical. Focus on data, patterns, and systematic approaches.');
      break;
    case 'Challenging':
      instructions.push('Be challenging and push back constructively. Ask tough questions to promote growth.');
      break;
    case 'Supportive':
    default:
      instructions.push('Be supportive and understanding. Prioritize emotional support and validation.');
  }
  
  // Focus areas
  if (preferences.focusAreas && preferences.focusAreas.length > 0) {
    instructions.push(`Focus coaching on: ${preferences.focusAreas.join(', ')}.`);
  }
  
  // Message length
  switch (preferences.messageLength) {
    case 'Brief':
      instructions.push('Keep responses very brief (1-2 sentences max).');
      break;
    case 'Detailed':
      instructions.push('Provide detailed responses with examples when helpful.');
      break;
    case 'Concise':
    default:
      instructions.push('Keep responses concise (2-4 sentences).');
  }
  
  // Emojis
  if (preferences.useEmojis) {
    instructions.push('Feel free to use emojis occasionally to add warmth.');
  } else {
    instructions.push('Avoid using emojis.');
  }
  
  return instructions.join(' ');
}

// Export both router and helper
module.exports = router;
module.exports.generatePromptInstructions = generatePromptInstructions;


